### PR TITLE
Qutip operators

### PR DESCRIPTION
**Pull request by araza6**
Further to https://github.com/qgrad/qgrad/issues/21, Pauli - X, Y, Z and creation and annihilation operators added. PR is still under works, for the remaining operators. Marking draft for now

### CODE REVIEW

**Comment by nathanshammah** [comment]
Please merge the current upstream master into this branch and push to origin, so we can check the tests with pytest in CI.

**Comment by nathanshammah** [comment]
CI is failing:
```
ERROR qgrad/tests/test_qutip.py::test_expect_sigmaxyz
```
Any updates on this, is it review ready after this is fixed? 

You could add `*.txt` or the txt file file itself to the gitignore to avoid having it being picked up in git.

**Comment by araza6** [comment]
@nathanshammah at the moment, 9 tests failing 13 passing. Should be review ready soon!

**Comment by araza6** [comment]
This is strange. Pytest passed all 23 tests on my machine but failed 1 here

File: .gitignore [hunk 1]
```diff
@@ -39,6 +39,8 @@ wheels/
 *.egg
 MANIFEST
 
+# txt files
```
Comment (nathanshammah) [inline .gitignore:42]:
```suggestion
*.swp
# txt files
```

File: qgrad/tests/test_qutip.py [hunk 2]
```diff
@@ -0,0 +1,358 @@
...
+        sigmay(), jnp.array([[0.0 + 0.0j, 0.0 - 1.0j], [0.0 + 1.0j, 0.0 + 0.0j]])
+    )
+
+
+def test_sigmaz():
+    assert_array_equal(sigmaz(), jnp.array([[1.0, 0.0], [0.0, -1.0]]))
+
+
+@pytest.mark.paramterize("op", [sigmax(), sigmay(), sigmaz()])
+@pytest.mark.paramterize("state", [basis(2, 0), basis(2, 1)])
```
Comment (nathanshammah) [inline qgrad/tests/test_qutip.py:170]:
```suggestion
@pytest.mark.parametrize("op", [sigmax(), sigmay(), sigmaz()])
@pytest.mark.parametrize("state", [basis(2, 0), basis(2, 1)])
```

File: qgrad/tests/test_qutip.py [hunk 3]
```diff
@@ -0,0 +1,358 @@
...
+    if op != sigmaz():
+        assert expect(op, state) == 0.0
+    elif state == basis(2, 0):
+        assert expect(op, state) == 1.0
+    else:
+        assert expect(op, state) == -1.0
+
+
+displace = Displace(2) #Initializing displace obj for `test_expect_herm`
+@pytest.mark.paramterize("oper, state", [(rand_herm(2).full(), basis(2, 0)),
```
Comment (nathanshammah) [inline qgrad/tests/test_qutip.py:183]:
```suggestion
@pytest.mark.parametrize("oper, state", [(rand_herm(2).full(), basis(2, 0)),
```

**Comment by araza6** [comment]
@nathanshammah thanks for quick feedback! The PR is review-ready now (cc @quantshah). The warning with JAX always shows up with CPUs. For the other pytest warnings, I am not sure if they are common.

**Comment by nathanshammah** [comment]
@araza6 can you request review when the PR is ready? (do you have access to the option, in the top right?) This will speed up review by @quantshah and me.

**Comment by araza6** [comment]
Hi @nathanshammah, PR is ready since the last commit, and you and @quantshah are already being shown as reviewers

**Comment by nathanshammah** [comment]
@araza6: You did not ask for review again at the end of these commits, this is why I am asking. By the way @quantshah and I are put as reviewers because I self-requested my review and that of Shahnawaz the other day, after some changes were made. You should ask again for review once you end up adding commits. This is good practice so that reviewers do not end up reviewing some code that is not ready or that is being locally modified, so that their time is not wasted. It is necessary to request again the review every time changes are added, so that a PR can have multiple back and forth, just like a paper between referees and authors.

**Comment by araza6** [comment]
@nathanshammah I see. I will take note of this

File: qgrad/qgrad_qutip.py [hunk 4]
```diff
@@ -0,0 +1,398 @@
...
+        float: fidelity between the two density matrices 
+    """
+    dm1, dm2 = jnp.asarray(a), jnp.asarray(b)
+    fidel = jnp.trace(sqrtm(jnp.dot(jnp.dot(sqrtm(dm1), dm2), sqrtm(dm1)))) ** 2
+    return jnp.real(fidel)
+
+
+def rot(params):
+    """
+    Returns a unitary matrix describing rotation around Z-Y-Z axis for a single qubit
```
Comment (nathanshammah) [inline qgrad/qgrad_qutip.py:70]:
```suggestion
    Returns a unitary matrix describing rotation around Z-Y-Z axis for a single qubit.
```

File: qgrad/qgrad_qutip.py [hunk 5]
```diff
@@ -0,0 +1,398 @@
...
+    if jnp.asarray(state).shape[1] >= 2:
+        return _expect_dm(oper, state)
+
+    else:
+        return _expect_ket(oper, state)
+
+
+def _expect_dm(oper, state):
+    """Private function to calculate the expectaion value of 
+    and operator with respect to a density matrix
```
Comment (nathanshammah) [inline qgrad/qgrad_qutip.py:227]:
```suggestion
    """Private function to calculate the expectation value of 
    an operator with respect to a density matrix
```

File: qgrad/qgrad_qutip.py [hunk 6]
```diff
@@ -0,0 +1,398 @@
...
+    """
+    # convert to jax.numpy arrays in case user gives raw numpy
+    oper, rho = jnp.asarray(oper), jnp.asarray(state)
+    # Tr(rho*op)
+    return jnp.trace(jnp.dot(rho, oper))
+
+
+def _expect_ket(oper, state):
+    """Private function to calculate the expectaion value of 
+    and operator with respect to a ket
```
Comment (nathanshammah) [inline qgrad/qgrad_qutip.py:237]:
```suggestion
    """Private function to calculate the expectation value of 
    an operator with respect to a ket.
```

File: qgrad/qgrad_qutip.py [hunk 7]
```diff
@@ -0,0 +1,398 @@
...
+        # Diagonal of the transformation matrix P, and apply to eigenvectors.
+        transform = self.t_scale * (alpha / np.abs(alpha)) ** -self.range
+        evecs = transform[:, None] * self.evecs
+        # Get the exponentiated diagonal.
+        diag = np.exp(1j * np.abs(alpha) * self.evals)
+        return np.conj(evecs) @ (diag[:, None] * evecs.T)
+
+
+def squeeze(N, z):
+    """Single-mode squeezing operator
```
Comment (nathanshammah) [inline qgrad/qgrad_qutip.py:272]:
```suggestion
    """Single-mode squeezing operator.
```

**Pull request merged**